tests/integration/backward_compatible/sql_test.py

```diff
@@ -0,0 +1,590 @@
+            result.iterator().result()
+
+    # Can't test the case we would expect row to be not set, because the IMDG SQL
+    # engine does not support update/insert queries now.
+    def test_is_row_set_when_row_is_set(self):
+        self._populate_map()
+        result = self.client.sql.execute("SELECT * FROM %s" % self.map_name)
+        self.assertTrue(result.is_row_set().result())
+
+    # Can't test the case we would expect a non-negative updated count, because the IMDG SQL
+    # engine does not support update/insert queries now.
+    def test_update_count_when_there_is_no_update(self):
+        self._populate_map()
+        result = self.client.sql.execute("SELECT * FROM %s WHERE __key > 5" % self.map_name)
+        self.assertEqual(-1, result.update_count().result())
+
+    def test_get_row_metadata(self):
+        self._populate_map(value_factory=str)
+        result = self.client.sql.execute("SELECT * FROM %s" % self.map_name)
+        row_metadata = result.get_row_metadata().result()
+        self.assertEqual(2, row_metadata.column_count)
+        columns = row_metadata.columns
+        self.assertEqual(SqlColumnType.INTEGER, columns[0].type)
+        self.assertEqual(SqlColumnType.VARCHAR, columns[1].type)
+        self.assertTrue(columns[0].nullable)
+        self.assertTrue(columns[1].nullable)
+
+    def test_close_after_query_execution(self):
+        self._populate_map()
+        result = self.client.sql.execute("SELECT * FROM %s" % self.map_name)
+        for _ in result:
+            pass
+        self.assertIsNone(result.close().result())
+
+    def test_close_when_query_is_active(self):
+        self._populate_map()
+        statement = SqlStatement("SELECT * FROM %s " % self.map_name)
+        statement.cursor_buffer_size = 1  # Each page will contain 1 row
+        result = self.client.sql.execute_statement(statement)
+
+        # Fetch couple of pages
+        iterator = iter(result)
+        next(iterator)
+
+        self.assertIsNone(result.close().result())
+
+        with self.assertRaises(HazelcastSqlError):
+            # Next fetch requests should fail
+            next(iterator)
+
+    def test_with_statement(self):
+        self._populate_map()
+        with self.client.sql.execute("SELECT this FROM %s" % self.map_name) as result:
+            six.assertCountEqual(
+                self, [i for i in range(10)], [row.get_object_with_index(0) for row in result]
+            )
+
+    def test_with_statement_when_iteration_throws(self):
+        self._populate_map()
+        statement = SqlStatement("SELECT this FROM %s" % self.map_name)
+        statement.cursor_buffer_size = 1  # so that it doesn't close immediately
+
+        with self.assertRaises(RuntimeError):
+            with self.client.sql.execute_statement(statement) as result:
+                for _ in result:
+                    raise RuntimeError("expected")
+
+        self.assertIsInstance(result.close(), ImmediateFuture)
```

Comment:
Is it hard in python to add a spy to close() and assert that it is called? I know it is easy in Node.js

Reply:
I have to use mock objects and we don't do it in integration tests. possible, but not that good looking. we already test lots of cases in unit tests so I am planning to not do it here

Reply:
Similarly spy should require mocking, you can spy a method without changing its behaviour in Node.js